File: src/app/services/fund-details.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { FundDetails } from '../models/fund-details.model';

@Injectable({
  providedIn: 'root'
})
export class FundDetailsService {
  private baseUrl = '/api/fund-details';

  constructor(private http: HttpClient) {}

  getApportionmentPercentages(): Observable<number[]> {
    return this.http.get<number[]>(`${this.baseUrl}/apportionment-percentages`);
  }

  performActionsBasedOnTotal(totalPercentage: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/perform-actions`, { totalPercentage });
  }

  deleteFundById(fundId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${fundId}`);
  }

  getFunds(): Observable<FundDetails[]> {
    return this.http.get<FundDetails[]>(this.baseUrl);
  }

  updateFund(fundId: string, apportionmentPercentage: number): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${fundId}`, { apportionmentPercentage });
  }

  deleteFund(fundId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${fundId}`);
  }

  fetchAvailableFunds(productCode: string, saDateRecd: string): Observable<FundDetails[]> {
    return this.http.get<FundDetails[]>(`${this.baseUrl}/available-funds`, {
      params: { productCode, saDateRecd }
    });
  }

  validateApportionment(percentage: number): Observable<boolean> {
    return this.http.post<boolean>(`${this.baseUrl}/validate-apportionment`, { percentage });
  }

  deleteFundRecord(recordId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/record/${recordId}`);
  }

  fetchImageData(currentBlock: string, currentItem: string, proposalType: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/image-data`, {
      params: { currentBlock, currentItem, proposalType }
    });
  }

  validateCorrectValue(correctValue: any): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/validate-correct-value`, { correctValue });
  }

  updateFundDetails(fundDetails: FundDetails): Observable<any> {
    return this.http.put<any>(this.baseUrl, fundDetails);
  }

  resetFundDetails(): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/reset`, {});
  }

  navigateToNextItem(): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/navigate-next`, {});
  }

  fetchFundNames(productCode: string, coverCode: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/fund-names`, {
      params: { productCode, coverCode }
    });
  }

  deleteRecords(): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/delete-records`);
  }

  getFundLOV(productCode: string, coverCode: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/fund-lov`, {
      params: { productCode, coverCode }
    });
  }

  fetchFundDetails(productCode: string, dateRange: string): Observable<FundDetails[]> {
    return this.http.get<FundDetails[]>(`${this.baseUrl}/details`, {
      params: { productCode, dateRange }
    });
  }

  addFundDetail(fundID: string, apportionmentPercentage: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/add`, { fundID, apportionmentPercentage });
  }

  updateFundDetail(fundID: string, apportionmentPercentage: number): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/update`, { fundID, apportionmentPercentage });
  }

  deleteFundDetail(fundID: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/delete/${fundID}`);
  }

  clearFields(): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/clear-fields`, {});
  }

  navigateToFirstRecord(): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/navigate-first`, {});
  }

  deleteFundRecords(): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/delete-fund-records`);
  }

  getSSOFundDetails(productCode: string, dateRange: string): Observable<FundDetails[]> {
    return this.http.get<FundDetails[]>(`${this.baseUrl}/sso-fund-details`, {
      params: { productCode, dateRange }
    });
  }

  deleteSSOFundRecord(recordId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/sso-fund/${recordId}`);
  }

  manageSolutionProductDetails(solutionProductDetails: any): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/solution-product`, solutionProductDetails);
  }

  getFundList(): Observable<FundDetails[]> {
    return this.http.get<FundDetails[]>(`${this.baseUrl}/fund-list`);
  }

  saveApportionment(fundId: string, percentage: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/save-apportionment`, { fundId, percentage });
  }

  validatePercentage(percentage: number): Observable<boolean> {
    return this.http.post<boolean>(`${this.baseUrl}/validate-percentage`, { percentage });
  }

  calculateTotalPercentage(percentages: number[]): Observable<number> {
    return this.http.post<number>(`${this.baseUrl}/calculate-total`, { percentages });
  }

  handleNavigation(totalPercentage: number, productCode: number, productDefinition: string): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/handle-navigation`, { totalPercentage, productCode, productDefinition });
  }
}

File: src/main/java/com/azbj/proposalde3/service/FundDetailsService.java
package com.azbj.proposalde3.service;

import com.azbj.proposalde3.repository.FundDetailsRepository;
import com.azbj.proposalde3.model.FundDetails;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
public class FundDetailsService {

    @Autowired
    private FundDetailsRepository fundDetailsRepository;

    public List<Double> getApportionmentPercentages() {
        List<FundDetails> fundDetailsList = fundDetailsRepository.findAll();
        return fundDetailsList.stream().map(FundDetails::getApportionmentPercentage).toList();
    }

    public void performActionsBasedOnTotal(double totalPercentage) {
        if (totalPercentage == 100) {
            // Enable specific fields based on product code and product definition
            // Navigate to the appropriate block and item
        } else {
            // Navigate to the next record and set focus on a specific field
        }
    }

    public void deleteFundById(String fundId) {
        if (fundDetailsRepository.existsById(fundId)) {
            fundDetailsRepository.deleteById(fundId);
        } else {
            throw new RuntimeException("Fund not found");
        }
    }

    public List<FundDetails> getFunds() {
        return fundDetailsRepository.findAll();
    }

    public void updateFund(String fundId, double apportionmentPercentage) {
        FundDetails fundDetails = fundDetailsRepository.findById(fundId).orElseThrow(() -> new RuntimeException("Fund not found"));
        fundDetails.setApportionmentPercentage(apportionmentPercentage);
        fundDetailsRepository.save(fundDetails);
    }

    public void deleteFund(String fundId) {
        if (fundDetailsRepository.count() > 1) {
            fundDetailsRepository.deleteById(fundId);
        } else {
            throw new RuntimeException("At least one fund is required");
        }
    }

    public List<FundDetails> fetchAvailableFunds(String productCode, String saDateRecd) {
        return fundDetailsRepository.findAvailableFunds(productCode, saDateRecd);
    }

    public void validateApportionment(double percentage) {
        if (percentage < 0 || percentage > 100) {
            throw new RuntimeException("Apportionment percentage should be between 0 and 100");
        }
    }

    public void resetFundDetails() {
        // Clear all fields related to fund details and set specific fields to null
    }

    public void navigateToNext() {
        // Navigate to the next item based on conditions
    }

    public void fetchImageData(String currentBlock, String currentItem, String proposalType) {
        // Fetch the correct image data based on the current block, item, and proposal type
    }

    public void validateCorrectValue(int correctValue) {
        // Validate the correct value against the database
    }

    public void manageSolutionProductDetails(Object solutionProductDetails) {
        // Handle the business logic for managing solution product details
    }

    public void fetchFundDetails(String productCode, String dateRange) {
        // Fetch fund details from the database based on the product code and date range
    }

    public void processApportionment(double apportionmentPercentage, String fundId, String fundName) {
        validateApportionment(apportionmentPercentage);
        FundDetails fundDetails = fundDetailsRepository.findById(fundId).orElseThrow(() -> new RuntimeException("Fund not found"));
        fundDetails.setApportionmentPercentage(apportionmentPercentage);
        fundDetails.setFundName(fundName);
        fundDetailsRepository.save(fundDetails);
    }

    public void fetchFundLOV(String productCode, String coverCode) {
        // Fetch the list of fund names based on the product code and cover code
    }

    public void deleteRecords() {
        // Delete records based on the validation flag
    }

    public void deleteFundRecords() {
        // Delete fund records based on the validation flag
    }

    public void fetchFundNames(String productCode, String coverCode) {
        // Fetch the fund names based on the product code and cover code
    }

    public void fetchFundDetails(String currentBlock, String currentItem) {
        // Fetch the fund details based on the current block and item
    }

    public void fetchFundDetails(String portfolioStrategy, String productDefinition, String productCode, String frequency, String packageCode) {
        // Fetch fund details based on the input parameters
    }

    public void updateFundDetails(FundDetails fundDetails) {
        fundDetailsRepository.save(fundDetails);
    }

    public void fetchFundDetails(String fundName) {
        // Fetch the fund details based on the fund name
    }

    public void fetchFundDetails(String productCode, String dateRange) {
        // Fetch the fund details based on the product code and date range
    }

    public void validateApportionmentPercentage(double percentage) {
        if (percentage < 0 || percentage > 100) {
            throw new RuntimeException("Apportionment percentage should be between 0 and 100");
        }
    }

    public double sumApportionmentPercentages() {
        List<FundDetails> fundDetailsList = fundDetailsRepository.findAll();
        return fundDetailsList.stream().mapToDouble(FundDetails::getApportionmentPercentage).sum();
    }

    public void navigateToBlock(double totalPercentage) {
        if (totalPercentage == 100) {
            // Navigate to the appropriate block and item
        } else {
            // Navigate to the next item in the current block
        }
    }
}
